### ISSUE TITLE
`esbuild` doesn't support `tsc`'s `outFile` style of builds

### ISSUE THREAD

**Issue by LoganDark** [evanw/esbuild#3100]
`tsconfig.json` supports an `outFile` option that will cause `tsc` to create a single output file from multiple input files. This is done by simply concatenating them together, and using triple-slash directives to declare dependencies. For example, say we have some TypeScript source files:

###### `a.ts`

```ts
const x = 5;
```

###### `b.ts`

```ts
/// <reference path="a.ts" />
console.log(x)
```

###### `c.ts`

```ts
/// <reference path="a.ts" />
const y = x + 5
```

If we run `tsc` on these files with an `outFile` set, it will generate one of these two output files:

###### `c.ts` first

```js
const x = 5
const y = x + 5
console.log(x)
```

###### `b.ts` first

```js
const x = 5
console.log(x)
const y = x + 5
```

Both of them satisfy the dependency of `b.ts` and `c.ts` on `a.ts`, but there is no real entry point here, nor are there any imports at runtime. Every file simply runs in sequence, in the same global scope. This is a different approach than bundling, which creates an isolated module and scope for each file, and requires them to explicitly import each other, starting from a single entrypoint.

We need to build this type of script for our project, where everything is at the top level and isn't in self-contained modules. The final script runs in the browser, and we don't want isolated or self-contained modules for it.

But right now, when we pass multiple "entrypoints" to esbuild and specify an `--outfile`, it complains that we have to use an `--outdir` instead. It doesn't seem to support this approach, nor does it care at all about the presence of triple-slash directives, even if we run it in non-bundler mode on one of the files as an "entrypoint".

It looks like something that `esbuild` could possibly support, though.

-Emily

**Comment by hyrious** [comment]
You're asking esbuild to add a new output format (in parallel with `esm`, `cjs`, `iife`) which is `concat`. This is out of esbuild's design now -- see #1700. Note that both rollup and webpack does not have this kind of build.

This feature is also hard to implement since esbuild doesn't know the executing order of the files, esbuild doesn't have a type system and it does not track unexported variable names across files.

**Comment by LoganDark** [comment]
> esbuild doesn't know the executing order of the files

The executing order is defined by the triple-slash directives; the post contains a basic example.

> esbuild doesn't have a type system

It doesn't need to, types do not matter. The files are compiled in complete isolation.

> it does not track unexported variable names across files

It doesn't need to, this is not based on variable usage whatsoever.

> Note that both rollup and webpack does not have this kind of build.

Because `tsc` has it on its own, and also because rollup and webpack are explicitly module based.

esbuild, however, has a non-bundler mode for compiling *scripts* that are not made of modules.

-Emily

**Comment by evanw** [comment]
It's true that esbuild doesn't support this. But that's ok because that's not something that a bundler needs to support, and esbuild is a bundler. I'm closing this issue because I consider this to be out of scope.

> But esbuild is designed to be used _in place of `tsc`_.

This is not true, sorry. I'm the one who designed esbuild and I did not design it as a replacement for `tsc`. I recommend that you use other tools for this instead.
